### PR TITLE
Add grapheme cluster support to String

### PR DESCRIPTION
**Pull request by naqvis**
Implements Unicode Text Segmentation according to Unicode Standard Annex #29 (Unicode version 13.0.0) to determine the grapheme cluster boundaries of unicode text.

Moving [naqvis/uni_text_seg](https://github.com/naqvis/uni_text_seg) shard to standard library.

https://forum.crystal-lang.org/t/unicode-text-segmentation-grapheme-clusters/3274

### CODE REVIEW

File: src/string/grapheme/grapheme.cr [hunk 1]
```diff
@@ -0,0 +1,226 @@
+require "./properties"
+
+module String::Grapheme
+  # Unicode extended grapheme cluster. `Graphemes` Iterator will return an instance of this struct to return information about specific grapheme cluster.
+  struct Cluster
```
Comment (asterite) [inline src/string/grapheme/grapheme.cr:5]: This is a great start, thank you! 🙏 

I was thinking... most of the time a cluster will be just a single char, right? I was thinking that a `Cluster` could still be a struct, but the underlying representation could be a union:
- if the cluster is simple (just one codepoint) then it's just that codepoint
- if the cluster consists of multiple codepoints, it's an array of codepoints. Or if clusters of two codepoints are common, we could also consider a `Tuple(Int32, Int32)` to be part another representation (same if three codepoints are common)
- we should probably not keep a reference to the original string. Imagine reading a huge string from a file and keeping the first grapheme. The grapheme will point back to the string so it will never be GCed. So in my mind a grapheme must be an independent entity.

File: src/string/grapheme.cr [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+  # code points into clusters perceived by the user as one character.
+  # ```
+  # "🔮👍🏼!".each do |cluster|
+  #   pp cluster.codepoints
+  #   pp cluster.positions
+  #   pp cluster.str
+  #   pp cluster.bytes
+  # end
+  # ```
+  def graphemes
```
Comment (asterite) [inline src/string/grapheme.cr:16]: I think this should return `Array(Grapheme::Cluster)`. Then we can have methods like `each_grapheme` that iterates over the graphemes or return an `Iterator` if there's no block. That would make it consistent with other String APIs.

File: src/string/grapheme/grapheme.cr (hunk 1 above)
Comment (naqvis) [reply src/string/grapheme/grapheme.cr:5]: Thank you Ary 🙏 for review and comments.

>     * if the cluster is simple (just one codepoint) then it's just that codepoint
Yes, there would be cases where cluster will have just one codepoint, but some emojis or some unicode chars can have 4 or more codepoints. for example `🧙‍♂️` will have 4 codepoints of `[129497, 8205, 9794, 65039]`.


>     * if the cluster consists of multiple codepoints, it's an array of codepoints. Or if clusters of two codepoints are common, we could also consider a `Tuple(Int32, Int32)` to be part another representation (same if three codepoints are common)

As for item 1, there could be clusters with multiple codepoints, so don't know what would be the optimal case for Tuple size? should we think of cases with 2,3,4 and so on?


>     * we should probably not keep a reference to the original string. Imagine reading a huge string from a file and keeping the first grapheme. The grapheme will point back to the string so it will never be GCed. So in my mind a grapheme must be an independent entity.

Code is not keeping reference to original string, but its keeping the codepoints and byte location of chars, and `Graphemes` constructor is the only place, where passed string is handled and then forget of. Please correct me if some portion of code is keeping the reference to original string.

File: src/string/grapheme.cr (hunk 2 above)
Comment (naqvis) [reply src/string/grapheme.cr:16]: Thank you Ary. Will make changes

Comment (naqvis) [reply src/string/grapheme.cr]: Made changes. Added `each_grapheme` methods as well

File: src/string/grapheme/grapheme.cr [hunk 3]
```diff
@@ -0,0 +1,226 @@
...
+        # be the one that just ended.
+        if @pos - 1 == 0 || boundary
+          @end = @pos - 1
+          break
+        end
+      end
+      @start != @end
+    end
+
+    # State::apheme cluster parser states
```
Comment (Fryguy) [inline src/string/grapheme/grapheme.cr:137]: Typo?

File: src/string/grapheme/grapheme.cr [hunk 4]
```diff
@@ -0,0 +1,226 @@
...
+      LVV
+      LVTT
+      Prepend
+      ExtendedPictographic
+      ExtendedPictographicZWJ
+      RIOdd
+      RIEven
+    end
+
+    # State::apheme cluster parser's breaking instructions.
```
Comment (Fryguy) [inline src/string/grapheme/grapheme.cr:152]: Also typo?

File: src/string/grapheme/properties.cr [hunk 5]
```diff
@@ -0,0 +1,1670 @@
...
+  end
+
+  # Maps code point ranges to their properties. In the context of this package,
+  # any code point that is not contained may map to "prAny". The code point
+  # ranges in this slice are numerically sorted.
+  #
+  # These ranges were taken from
+  # http://www.unicode.org/Public/UCD/latest/ucd/auxiliary/GraphemeBreakProperty.txt
+  # as well as
+  # https://unicode.org/Public/emoji/latest/emoji-data.txt
```
Comment (Fryguy) [inline src/string/grapheme/properties.cr:50]: This link doesn't seem to work.  Maybe one of these? https://unicode.org/Public/emoji/13.1/

File: src/string/grapheme/properties.cr [hunk 6]
```diff
@@ -0,0 +1,1670 @@
...
+  # ("Extended_Pictographic" only) on March 11, 2019. See
+  # https://www.unicode.org/license.html for the Unicode license agreement.
+  @@codepoints : Array(Tuple(Int32, Int32, Property))?
+
+  # :nodoc:
+  protected def self.codepoints
+    @@codepoints ||= begin
+      data = Array(Tuple(Int32, Int32, Property)).new
+
+      put(data, {0x0000, 0x0009, Property::Control})                # Cc  [10] <control-0000>..<control-0009>
```
Comment (Fryguy) [inline src/string/grapheme/properties.cr:60]: Would it be possible to autogenerate this list from the actual file as provided by unicode.org, perhaps using a macro?  My concern is that this will be difficult to keep up to date as new emoji versions come out, but if we can automate it, then generating it is as easy as downloading the latest text file.

File: src/string/grapheme.cr [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+
+  # Yields each Unicode extended grapheme cluster in the string to the block.
+  #
+  # ```
+  # "🧙‍♂️💈".each_grapheme do |cluster|
+  #   p! cluster.codepoints
+  #   p! cluster.to_s
+  # end
+  # ```
+  def each_grapheme : Nil
```
Comment (Blacksmoke16) [inline src/string/grapheme.cr:23]:
```suggestion
  def each_grapheme(& : String::Grapheme::Cluster -> Nil) : Nil
```

File: src/string/grapheme.cr [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+  # returns graphemes cluster iterator over Unicode extended grapheme clusters.
+  # ```
+  # "🔮👍🏼!".each do |cluster|
+  #   pp cluster.codepoints
+  #   pp cluster.positions
+  #   pp cluster.str
+  #   pp cluster.bytes
+  # end
+  # ```
+  def each_grapheme
```
Comment (Blacksmoke16) [inline src/string/grapheme.cr:38]:
```suggestion
  def each_grapheme : Iterator(String::Grapheme::Cluster)
```

File: src/string/grapheme/properties.cr (hunk 5 above)
Comment (Fryguy) [reply src/string/grapheme/properties.cr:50]: Or maybe https://unicode.org/Public/13.0.0/ucd/emoji/emoji-data.txt

File: src/string/grapheme.cr [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+
+class String
+  # returns an array of all Unicode extended grapheme clusters, specified in the Unicode Standard Annex #29. Grapheme clusters correspond to
+  # "user-perceived characters". These characters often consist of multiple code points (e.g. the "woman kissing woman" emoji consists of 8 code points:
+  # woman + ZWJ + heavy black heart (2 code points) + ZWJ + kiss mark + ZWJ + woman) and the rules described in Annex #29 must be applied to group those
+  # code points into clusters perceived by the user as one character.
+  # ```
+  # "🧙‍♂️💈".graphemes # => [String::Grapheme::Cluster(@codepoints=[129497, 8205, 9794, 65039], @positions={0, 13}), String::Grapheme::Cluster(@codepoints=[128136], @positions={13, 17})]
+  # ```
+  def graphemes
```
Comment (Blacksmoke16) [inline src/string/grapheme.cr:11]:
```suggestion
  def graphemes : Array(String::Grapheme::Cluster)
```

**Comment by straight-shoota** [comment]
Thanks for all review comments! However, I'd ask to restrict discussion on the general API design first. We can worry about type annotations, typos and implementation details later in the process of polishing up this PR. ✌️

**Comment by Fryguy** [comment]
Ah sorry!  I assumed it was already discussed ahead of time.

**Comment by naqvis** [comment]
Thank you all for reviews and suggestions. I've already (but haven't pushed yet to this PR)
* Fixed Typos
* Added Type Annotations
* Added script to auto-generate the `Properties` data

I will wait for further review and suggestions on API design, before I proceed with pushing above cited changes.

For people interested to see or use, you can see the above changes to original shard [naqvis/uni_text_seg](https://github.com/naqvis/uni_text_seg)

**Comment by straight-shoota** [comment]
This is really great, thanks for adding it to the stdlib.

We should probably dive a bit deeper into an optimal representation of a grapheme cluster. This is tightly related to how the API is used. So I'm not entirely sure about that.
I would expect that typical use cases treat single graphemes as a character-oriented value. Then `String` might actually be a more appropriate representation than an array of codepoints. [Julia's Unicode implementation](https://docs.julialang.org/en/v1/stdlib/Unicode/#Unicode.graphemes) and the Rust crate [unicode-segmentation](https://unicode-rs.github.io/unicode-segmentation/unicode_segmentation/index.html) represent graphemes as strings, for example.

Considering that in many typical use cases the vast majority of graphemes consists of just a single codepoint, this offers room for optimization. A single-codepoint grapheme could be represented as a `Char`, thus creating a union `String | Char` (as originally suggested in https://github.com/crystal-lang/crystal/pull/10721/files#r634603666).
A different optimization strategy could be to use a string pool to represent common graphemes. I think I like that more, because it's entirely abstracted away and does not add complexity to the API. Still just a rough idea, though (the good part of this optimization is that it can be delayed as it does not affect the API).

I'm also questioning the dedicated `Cluster` type itself. It combines the grapheme representation with the position in the original string. I don't think this is necessary. When iterating the characters of a string (`#each_char`), the position is not part of the character. If you need the position, you can use `#each_char_with_index`. I prefer this design for graphemes as well, such that a single grapheme has no connection to the original string. Then two equivalent graphemes from different original strings have identical memory representation.
So... if `Cluster` just wraps a single `String` property, there is little reason to actually have it as a separate type. We can just use `String` directly. Of course, a dedicated type adds a bit more type rigidity. But I don't think that's very useful. A grapheme cluster consists of a sequence of codepoints and the best data type to represent that is `String`. The only semantic difference between a string representing a grapheme and any other string is that the former definitely does not contain more than one grapheme.
A somewhat related comparison could be `String#each_line` it returns `String` and we can be sure that a string representing a line does not contain more than one line. But that does not warrant a separate data type `Line`.

So, combining that all, I propose the most basic interface should be `String#each_grapheme(& : String ->)` (plus extensions like iterator, `*_with_index` etc.).

**Comment by naqvis** [comment]
Thank you Straight-shoota and totally agree that there is a strong need of deeper dive into the design of API and i'll be more than pleased to participate in such discussions and/or suggestions. 

Just to clarify: initial intention of having `Cluster` struct was to provide more details to the user, current struct provides:

* codepoints
* string representation 
* byte indexes of grapheme in original string

>... and the Rust crate [unicode-segmentation](https://unicode-rs.github.io/unicode-segmentation/unicode_segmentation/index.html) represent graphemes as strings, for example.

Rust implementation also provides [grapheme_indices](https://unicode-rs.github.io/unicode-segmentation/unicode_segmentation/trait.UnicodeSegmentation.html#tymethod.grapheme_indices) which returns an iterator over grapheme cluster and their respective bytes offset.



> I'm also questioning the dedicated `Cluster` type itself. It combines the grapheme representation with the position in the original string. I don't think this is necessary. When iterating the characters of a string (`#each_char`), the position is not part of the character. If you need the position, you can use `#each_char_with_index`. I prefer this design for graphemes as well, such that a single grapheme has no connection to the original string. Then two equivalent graphemes from different original strings have identical memory representation.

Would quote [Unicode Anex#29](http://www.unicode.org/reports/tr29/#Grapheme_Cluster_Boundaries) to explain Grapheme cluster
> .. To avoid ambiguity with the computer use of the term character, this is called a user-perceived character. For example, “G” + grave-accent is a user-perceived character: users think of it as a single character, yet is actually represented by two Unicode code points. These user-perceived characters are approximated by what is called a grapheme cluster, which can be determined programmatically.  So we can't simply rely on `each_char` or `char_with_index` to determine grapheme cluster.

 Just an example: "🧙‍♂️💈" is two grapheme cluster but
```
"🧙‍♂️💈".chars # => ['🧙', '‍', '♂', '️', '💈']

# OR
"🧙‍♂️💈".each_char_with_index do |c,i|
  pp "c: #{c}, i: #{i}"
end
```

**Comment by straight-shoota** [comment]
> Rust implementation also provides grapheme_indices which returns an iterator over grapheme cluster and their respective bytes offset.

Yes, the equivalent would be the `#each_grapheme_with_index` method I mentioned.

I'm pretty sure I grasp the concept of graphemes vs. characters. I just mentioned `each_char` etc. for comparison, because it's actually pretty similar with regard to the relation between character/grapheme and the byte position. I.e. the `Char` type does not have a `position` property, because each `Char` instance is independent of any string. A type representing a grapheme IMO should be equally independent and not have a position property either.

**Comment by HertzDevil** [comment]
If we follow `String`'s conventions, `String::Grapheme::Graphemes` should mirror `Char::Reader`'s API and all access to positional information should be done through that type exclusively. Also it should be a `class`.

Is it possible to avoid iterating through the entire string on construction of a `Graphemes`?

**Comment by straight-shoota** [comment]
> Is it possible to avoid iterating through the entire string on construction of a Graphemes?

Yes, that should certainly work. But that's already digging into implementation details. Let's focus on the grand scheme for now.

**Comment by HertzDevil** [comment]
O(1) iterator construction is a natural non-functional requirement to ask for; if that's not possible, IMO the standard library should not offer an iterator API at all.

I looked at the code briefly and I believe `String::Grapheme::Graphemes` could itself depend on `Char::Reader`. Doing so also offers the possibility to start grapheme cluster iteration in the middle of a string.

**Comment by naqvis** [comment]
Based on previous discussions, comments and suggestions I'm thinking of something like

### Option 1  - Integrate with `String` 

#### Pros
* Natural to user of the String As it follows the same set of API which String offers for `Char`
* Make Graphemes a first class citizen of String
#### Cons
* Bloat String API

#### API
```
def graphemes : Array(Char | String)
def each_grapheme(& : (Char | String) -> Nil) : Nil
def each_grapheme : Iterator(Char | String)
def each_grapheme_with_index(offset = 0, & : {Char | String, Int32} -> Nil) : Nil
```

### Option 2  - Add `Char::GraphemeReader` 

#### Pros
* Don't bloat `String` API
* Separate API
#### Cons
* Make Graphemes  2nd class citizen as user would have to separately instantiate instance

#### API
```
API similar to Char::Reader https://crystal-lang.org/api/1.0.0/Char/Reader.html
```


Any thoughts?

**Comment by straight-shoota** [comment]
I'm not even sure both options are mutually exclusive. A similar API exists for characters as well: `#chars`, `#each_char`, `#each_char_with_index` (like Option 1). `CharReader` serves as a basis for those implementations, and it can also be used separately (like Option 2).

That being said, I definitely want the methods from Option 1. This is not bloating up the API, it's just adding useful features. And it does it in an approachable way.

Whether we need a separate `GraphemeReader` type may be up for debate. Sure, we're gonna need an iterator type, but that could be private with `Iterator(String)` as the external interface.
`CharReader` offers high flexibility for reading in a string (jumping around, going backwards etc.) which is mostly useful for parsing purposes. I don't think that's tremendously useful for graphemes. In any case, this could be added later if a need arises. So I'd go with Option 1 only for now.

**Comment by naqvis** [comment]
Thanks stratigh-shoota

> I'm not even sure both options are mutually exclusive. A similar API exists for characters as well: `#chars`, `#each_char`, `#each_char_with_index` (like Option 1). `CharReader` serves as a basis for those implementations, and it can also be used separately (like Option 2).

Yes true. But instead of having both APIs (String API using GraphemeReader), I was thinking of just having one set of API and have user use that when needed (so the choice of exclusive option was given). For sure we could have Option 3 (Option1 + Option2), but totally agree that having Char::Reader like api at this stage won't be much beneficial, but should the need arise in future, that could be added.

Now we have some kind of tentative understanding and/or how the API should be like, can we have also have a quick discussion on **eager** vs **lazy** processing of String for grapheme clusters?

>  we should probably not keep a reference to the original string. Imagine reading a huge string from a file and keeping the first grapheme. The grapheme will point back to the string so it will never be GCed. So in my mind a grapheme must be an independent entity.

Per Ary initial comment, he is against (or he recommends not to do so) keeping a reference to original string, so this will rule out **lazy** treatment of string. And this is how code submitted to this PR is handling that.

But HertzDevil is suggesting against **eager** treatment.

> Is it possible to avoid iterating through the entire string on construction of a `Graphemes`?

**Comment by naqvis** [comment]
> The iterator should definitely be lazy, there should be no doubt about that. All iterators work like that. This obviously means the iterator needs a reference to the original string. But once iteration is finished, it gets released. This approach allows much flexibility: If you want to get rid of the original string ASAP, you can collect the iterator into an array and the original resource can be released while iterating that.

Yes, and PR pushed implementation is providing the iterator (`Grapheme::Graphemes` is Iterator), and it is storing the code points of original string (instead of keeping the reference to original string) and perform lazy determination of graphemes on `iterator#next`. So pushed code is doing lazy determination, but question was on whether to keep reference to original string or not.

If we agree that `Grapheme::Graphemes` can keep a reference to original string, then class itself doesn't need to keep anything else other than reference to original string.

I will make changes and pushed modified code to this PR for further review and discussion.

**Comment by straight-shoota** [comment]
The current implementation in this PR is not lazy because it iterates the string on initialization. And the resulting intermediary array of codepoints occupies additional memory and even more than the original string.

> If we agree that `Grapheme::Graphemes` can keep a reference to original string, then class itself doesn't need to keep anything else other than reference to original string.

Yes, it definitely should do that. It should iterate the original string on iteration. And probably use a `CharReader` for that.

**Comment by naqvis** [comment]
> The current implementation in this PR is not lazy because it iterates the string on initialization. And the resulting intermediary array of codepoints occupies additional memory and even more than the original string.

My claim of lazy is **lazy determination of grapheme cluster** and yes current implementation does iterates through original string to store its code points as well as the byte indexes. This was done to avoid keeping the reference to original string. 

So keeping reference to original string will remove that portion of code.

**Comment by naqvis** [comment]
Pushed following changes to this PR.
* Added scripts to generate `Properties` data
* Added Type annotations to API
* Fixed Typos
* API in Option 1 

I've removed `each_grapheme_with_index` as it's confusing to user if we just increment the counter by 1 (follow the same `each_char_with_index` is doing), also I had thought of providing the `Char::Reader#pos` as the index, but found that might be confusing to its user.

**Comment by straight-shoota** [comment]
Thanks.

As mentioned above, I'm actually leaning more towards using only `String` to represent graphemes, even single-codepoint ones. We could use more comments on this, though.

**Comment by naqvis** [comment]
Thank you straight-shoota. I also prefer `String`, but switched to `Union` based on my understanding of Ary initial comments on union types 😄 

For sure will keep this open for further comments and/or suggestions.

**Comment by asterite** [comment]
Sorry, I didn't have more time to review this.

But to be clear, I don't think we should return `Char | String`. We should have a `struct Grapheme` that internally is represented as either a `Char` or a `String`. It has a `to_s(io : IO)` and so on, so if you don't need a string representation of the grapheme the underlying representation can be kept as `Char`.

In fact, a Grapheme can probably consist of one, two or multiple chars, so the underlying representation could even be `Char | {Char, Char} | String` if needed. It doesn't matter. What matters is not exposing the underlying representation, and instead always return an opaque `Grapheme` type.

**Comment by asterite** [comment]
Swift has a `Character` type that represents a grapheme: https://developer.apple.com/documentation/swift/character

Note that the underlying representation is never revealed.

**Comment by straight-shoota** [comment]
Yeah, if we use a separate data type to encapsulate the representation, an internal union type works well.

I kind of like the simplicity of just using `String` - no union, no special type. This approach is used in other implementations as mentioned before (https://github.com/crystal-lang/crystal/pull/10721#issuecomment-844250987). So it certainly works.

A `Grapheme` type promises a performance benefit because it requires way less memory allocations in typical applications. And I suppose that common use cases don't need the individual graphemes as isolated strings.

**Comment by asterite** [comment]
Yes, that was my reasoning too. If you have a string like "hello world" and you ask for its grapehems, you end up with one string per char, which is a waste of memory. I guess in most cases a grapheme will be exactly one Char and so on the example above there would be no memory al locations when traversing those graphemes.

**Comment by naqvis** [comment]
Thank you Ary and Straight-shoota

> What matters is not exposing the underlying representation, and instead always return an opaque `Grapheme` type.

The very first push were doing things in a way which Ary suggested. So I've added `Grapheme::Cluster` struct which will be returned and it has overriden `to_s(io: IO)` method which will write the `String` representation of cluster to IO.

**Comment by straight-shoota** [comment]
Okay, sounds good. Let's keep going with that.

I'm wondering, why is it called `Cluster`? IMO `Grapheme` would be a more obvious name.

Now, the point of that type is to wrap a `Char | String` value. That should be used in stead of an array of chars. As mentioned before, I think the `pos` property should go to make every instance independent of the original string.

Regarding the actual implementation: In my experience, it is much easier to reason about a yielding iteration method than an iterator class which spreads the entire algorithm between individual method calls and shared status. So I would urge you to focus on that first, i.e. an implementation of `each_grapheme(& : Grapheme::Cluster -> Nil)` not based on the iterator type. I think that could help a lot while going forward. The iterator can be left out for now. We'll get there later.

**Comment by naqvis** [comment]
> I'm wondering, why is it called `Cluster`? IMO `Grapheme` would be a more obvious name.

Unicode call them `Grapheme Cluster`. Since code is located under `String::Grapheme` namespace, so it make more sense to call it `Cluster` thus reading as `Grapheme::Cluster`.
 
> Now, the point of that type is to wrap a `Char | String` value. That should be used in stead of an array of chars. As mentioned before, I think the `pos` property should go to make every instance independent of the original string.

I don't understand why internal representation is causing an argument here? Or what extra benefit it will have to stick to `Char | String` for internal representation of identified grapheme clusters?  
 
> Regarding the actual implementation: In my experience, it is much easier to reason about a yielding iteration method than an iterator class which spreads the entire algorithm between individual method calls and shared status. So I would urge you to focus on that first, i.e. an implementation of `each_grapheme(& : Grapheme::Cluster -> Nil)` not based on the iterator type. I think that could help a lot while going forward. The iterator can be left out for now. We'll get there later.

In the very beginning it was suggested that, code should follow the existing String API and that's how `each_char` methods of `String` are implemented.

IMHO it will make much more sense if we can first get to a conclusion on how core team is expecting the implementation to be to meet the required standard . This will save core team and myself whole lot of effort.

**Comment by straight-shoota** [comment]
> I don't understand why internal representation is causing an argument here?

The internal representation is the entire point of having a dedicated data type. That allows to transparently use a union of `Char | String` for the purpose of avoiding a lot of unnecessary heap allocations. If that wasn't the intention, we wouldn't use a separate type, but just `String` to represent a grapheme.

> Or what extra benefit it will have to stick to `Char | String` for internal representation

The benefit is that `Char` does not allocate heap memory, thus it's way more efficient than `String` for single-codepoint graphemes (which is the vast majority in typical use cases).

Now, we certainly don't need to focus too much on the performance optimization for the first iteration, but at least the basics should be set. That means, the internal value should at least be `String`, not an array of characters. `Char` can be added later to the union type, but `String` is the most basic type to represent a grapheme consisting of an arbitrary number of codepoints.

> In the very beginning it was suggested that, code should follow the existing String API and that's how each_char methods of String are implemented.

`#each_char` is probably a bad role model for this. It's not based on an iterator, though. It uses `Char::Reader`, which is not an iterator and stack allocated.

I want to be clear: It's not *required* that you implement the yielding method in the manner I described. I would happily accept a solid implementation based on an iterator, and the yielding variant can be added later. The current iterator implementation is complex and hard to follow. I believe it would be much easier to understand and thus review the algorithm, if it was implemented in a yielding method. Thus, I *recommend* to follow my suggestion.

**Comment by naqvis** [comment]
Thanks Straight-shoota.

Made changes as below:

1. Changed internal represtation to `Char | String`
2. Removed `pos` getter
3. Moved Iterator to a separate class
4. Changed `each_grapheme` method implementation to use yielding iteration instead of relying on Iterator.
5. `each_grapheme` Iterator method returns the `Grapheme::GraphemeIterator` instead instead.

Please help to review and suggest what further changes need to be made.

**Comment by straight-shoota** [comment]
I believe there may be some communication difficulties. I'm not sure what I can do to help it, but I'll try to be more explicit. Let me now if you have any idea how I can improve that.

My suggestion was not just about avoiding a class that implements `Iterator`. I would like to have the entire iteration logic for `#each_grapheme` in the method itself, not iterating over a sequence of `next` calls. That's essentially the same as the iterator approach, just that it's not technically an `Iterator`.

I was thinking about something like this:

```cr
def each_grapheme(& : Grapheme -> Nil) : Nil
  reader = Char::Reader.new(self)
  state = Grapheme::State::Any
  last_char = reader.next_char
  last_boundary = 0
  reader.each do |char|
    state, boundary = Grapheme.process_transition(state, last_char, char)
    last_char = char

    if boundary
      index = reader.pos
      # using only String here for simplicity, the Char optimization can follow
      yield Grapheme.new(byte_slice(last_boundary, index - last_boundary))

      last_boundary = index
    end
  end

  yield Grapheme.new(byte_slice(last_boundary, bytesize - last_boundary))
end
```

(Disclaimer: This implementation is probably not production ready and requires some refinements, but it serves as illustration of the concept I have in mind.)

`process_transition` would implement the transition logic from `move_next`.

I think this code is more concise and easier to follow than the state-based `Graphemes` iterator. It helps a lot for understanding what's going on, when you don't already know how the algorithm works.
Do you see what I mean?

The iterator implementation can likewise rely on `process_transition`, and should become relatively lightweight by that.

**Comment by naqvis** [comment]
> I believe there may be some communication difficulties. I'm not sure what I can do to help it, but I'll try to be more explicit.

I don’t believe it’s communication issue but its more on you are expecting the code to meet your already perceived solution. 

That being said, I don’t believe two people can produce same code for same problem. So let’s stop this conversation here. I’ll rather have this PR closed or cancelled.

Appreciate all the reviews and suggestions.

**Comment by straight-shoota** [comment]
If you believe that's my expectation here, then I'm pretty sure there is communication issue. I don't expect the code to look like I would write it. I'm happy to accept any alternative solution. I would actually be happy about learning a new approach 🙇 

But I need to understand the code, before I can approve it. And I want others to be able to understand it. That means it needs to readable and of high quality.
I think that's hard to accomplish with the approach you proposed. That's why I suggested an alternative, which in my experience can lead to a better solution that's easier to understand. You don't have to follow that. My intention was just to provide helping advice to get this PR improved for being accepted.

I'm sorry if my comments lead you to the impression I'm imposing anything on you.
I understand I you don't want to continue on this. That's totally fine.
However, I would be happy if you did. I will keep out of this discussion unless asked for comment.

**Comment by naqvis** [comment]
Thank you Straight-shoota, but I don't believe it make more sense for me to continue with this PR. So I will take the liberty to have it closed with this final comment.

**Comment by beta-ziliani** [comment]
@naqvis, let me intercede even if late. I understand your frustration. To me, the distinction of where is the logic of the iteration is not essential: the majority of the logic goes in the unavoidable `move_next` function. What is really crucial is the API, and I think we are more or less on the same page there. The implementation details are something we can work with with time.

So, will it be possible for you to reconsider re-opening this PR? I won't ask you to any further changes to the code. If we find we need to fix something, then we'll take care of that ourselves. :pray:

**Comment by asterite** [comment]
> If we find we need to fix something, then we'll take care of that ourselves.

This is a great comment. In the Ruby community I've seen some PRs being merged, and immediately after the authors made changes to the code to match their style and wished. I believe we should start doing that, because it removes the burden off PR authors to keep doing more and more changes, many of which happen without a total consensus.

**Comment by beta-ziliani** [comment]
It'd be great to have [PRs over PRs](https://github.com/isaacs/github/issues/959). But there is [something similar](http://bentrengrove.com/blog/2020/7/8/how-to-stack-prs-in-github) we can try.

**Comment by naqvis** [comment]
Thank you @beta-ziliani and appreciate your comments. I'm not against code changes and/or refactoring,  and will be more than happy to make changes to meet the standard of stdlib.

**Comment by beta-ziliani** [comment]
Thanks a lot @naqvis ! I only have two nitpicks to note:

- Start each comment with Uppercase.
- Fix the doc in `each_grapheme : Iterator(...)`, which does not correspond to the function.

File: scripts/generate_grapheme_properties.cr [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+props = parse_graphemes_data(body)
+
+body = HTTP::Client.get("#{UCD_ROOT}13.0.0/ucd/emoji/emoji-data.txt").body
+props["ExtendedPictographic"] = parse_emoji(body)
+
+props_data = props.values.flatten.sort! { |a, b| a.low <=> b.low }
+
+output = String.build do |str|
+  ECR.embed "#{__DIR__}/grapheme_properties.ecr", str
+end
```
Comment (Sija) [inline scripts/generate_grapheme_properties.cr:81]:
```suggestion
output = ECR.render("#{__DIR__}/grapheme_properties.ecr")
```

File: src/string/grapheme.cr [hunk 11]
```diff
@@ -0,0 +1,42 @@
+require "./grapheme/grapheme"
+
+class String
+  # returns an array of all Unicode extended grapheme clusters, specified in the Unicode Standard Annex #29. Grapheme clusters correspond to
+  # "user-perceived characters". These characters often consist of multiple code points (e.g. the "woman kissing woman" emoji consists of 8 code points:
+  # woman + ZWJ + heavy black heart (2 code points) + ZWJ + kiss mark + ZWJ + woman) and the rules described in Annex #29 must be applied to group those
+  # code points into clusters perceived by the user as one character.
+  # ```
```
Comment (Sija) [inline src/string/grapheme.cr:8]: Please, use an empty line between the description and an example.
ditto elsewhere

File: scripts/generate_grapheme_properties.cr [hunk 12]
```diff
@@ -0,0 +1,83 @@
...
+    i += 1
+  end
+  arr
+end
+
+def parse_graphemes_data(body)
+  result = Hash(String, Array(RRange)).new
+  body.each_line do |line|
+    line = line.strip
+    next if line.empty?
```
[the remark this answers is not in the document]
Comment (naqvis) [reply scripts/generate_grapheme_properties.cr:33]: Thanks Sijia. Shouldn't it be `next unless line = line.strip.presence` ?

[the remark this answers is not in the document]
Comment (Sija) [reply scripts/generate_grapheme_properties.cr:33]: Of course you're correct, my bad :)

**Comment by naqvis** [comment]
Thank you  @beta-ziliani  and @Sija  for review and suggestions. I've made suggested changes and pushed code to this PR.

**Comment by docelic** [comment]
Hey where do we stand on this PR? Would love to see this happen and thanks for everyone's effort on it.

File: scripts/generate_grapheme_properties.cr [hunk 13]
```diff
@@ -0,0 +1,79 @@
...
+    next if f2 < 0xFF
+    emoji << RRange.new(f1, f2, "ExtendedPictographic")
+  end
+  shapeup(emoji)
+end
+
+body = HTTP::Client.get("#{UCD_ROOT}UCD/latest/ucd/auxiliary/GraphemeBreakProperty.txt").body
+props = parse_graphemes_data(body)
+
+body = HTTP::Client.get("#{UCD_ROOT}13.0.0/ucd/emoji/emoji-data.txt").body
```
Comment (HertzDevil) [inline scripts/generate_grapheme_properties.cr:72]: Is it possible not to hardcode a specific version of Unicode here?

File: scripts/grapheme_properties.ecr [hunk 14]
```diff
@@ -0,0 +1,64 @@
...
+  end
+
+  # Maps code point ranges to their properties. In the context of this package,
+  # any code point that is not contained may map to `Property::Any`. The code point
+  # ranges in this slice are numerically sorted.
+  #
+  # These ranges were taken from
+  # http://www.unicode.org/Public/UCD/latest/ucd/auxiliary/GraphemeBreakProperty.txt
+  # as well as
+  # http://www.unicode.org/Public/13.0.0/ucd/emoji/emoji-data.txt
```
Comment (HertzDevil) [inline scripts/grapheme_properties.ecr:45]: ditto

File: spec/std/string/grapheme_spec.cr [hunk 15]
```diff
@@ -0,0 +1,657 @@
...
+    {original: "👩‍❤️‍💋‍👩", expected: [[0x1f469, 0x200d, 0x2764, 0xfe0f, 0x200d, 0x1f48b, 0x200d, 0x1f469]]},
+    {original: "🏋🏽‍♀️", expected: [[0x1f3cb, 0x1f3fd, 0x200d, 0x2640, 0xfe0f]]},
+    {original: "🙂", expected: [[0x1f642]]},
+    {original: "🙂🙂", expected: [[0x1f642], [0x1f642]]},
+    {original: "🇩🇪", expected: [[0x1f1e9, 0x1f1ea]]},
+    {original: "🏳️‍🌈", expected: [[0x1f3f3, 0xfe0f, 0x200d, 0x1f308]]},
+
+    # The following tests are taken from
+    # http://www.unicode.org/Public/12.0.0/ucd/auxiliary/GraphemeBreakTest.txt,
+    # see https://www.unicode.org/license.html for the Unicode license agreement.
```
Comment (HertzDevil) [inline spec/std/string/grapheme_spec.cr:27]: The specs that follow should also be generated from an ECR file.

File: scripts/generate_grapheme_properties.cr (hunk 13 above)
Comment (Blacksmoke16) [reply scripts/generate_grapheme_properties.cr:72]: Would it make sense to integrate it into the existing unicode data script? Then there's just one script for both contexts.

**Comment by asterite** [comment]
Regarding names, it seems there's:

- `String::Grapheme`
- `String::Grapheme::Cluster`
- `String::Graphemes`
- `String::GraphemeIterator` (hidden)

I think it's a bit confusing that `Grapheme` is not a grapheme. Plus there's `Grapheme` which is publicly exposed but then only method it has is `next`. Then there's `GraphemeIterator` which is essentially a very thing wrapper around `Graphemes`.

I propose we use these names:
- `Grapheme`: replaces `String::Grapheme::Cluster` (note that this also goes into the top level... why a Grapheme is nested inside String? Are there other kind of graphemes out there?)

That's it! There's no need to expose `String::Graphemes` because its API is the same as the iterator's one.

File: scripts/grapheme_properties.ecr [hunk 16]
```diff
@@ -0,0 +1,64 @@
...
+  # ("Extended_Pictographic" only). See
+  # https://www.unicode.org/license.html for the Unicode license agreement.
+  @@codepoints : Array(Tuple(Int32, Int32, Property))?
+
+  # :nodoc:
+  protected def self.codepoints
+    @@codepoints||= begin
```
Comment (HertzDevil) [inline scripts/grapheme_properties.ecr:52]: This can be expressed as a `protected class_getter`.
